### PR TITLE
DEV-43744-alert-eval-api

### PR DESCRIPTION
**Pull request by yasmin-tr**
**What is this feature?**

Logzio Alert Evaluation API

* contains basic functionality for alert evaluation api
* including headers and necessary files and models to support evaluation on logzio datasources

**Explaining the main logic and changes:** 
We added a new internal api for evaluating alerts externally from the alert-manager.
External in this meaning is that the evaluation is called outside the alert-manager and not as part of the engine schedule as it is today. 
To be able to reuse the evaluation code as much as possible, we are using the ScheduleService code to run alert evaluation. 
In order to use ScheduleService, but without having it running the evaluation, I needed to add a configuration not to execute the evaluations from the scheduler. That way we have all the management of the rule routines and updates, but only the execution is disabled, and then can run from our API.
For running the external evaluation I also added another function for the Scheduler interface that the API is using.

External notification support -
In order to send the notification in an external API, added a LogzioRouter that implements the Send interface which sends the PostableAlerts to the url coming from configuration

Logzio datasources support required the following changes:
* adding headers that contains the user-context passed by the API
* propagating the headers into the evaluation request in relevant clients - elasticsearch and prometheus
* adding url-override from the evaluation request passed by the API
* overriding the url of the datasource for the relevant one so it will call the logzio APIs for query

**Why do we need this feature?**

metrics alerts evaluation process

**Who is this feature for?**

Logzio Metrics Alerts

**Which issue(s) does this PR fix?**:

DEV-43744
DEV-43883
DEV-43889

Added code from the following tickets:
* DEV-31493 Override datasource URL and pass custom headers to alert rule evaluator (partly)
* DEV-17927 - add LogzIoHeaders, Add error msg, change time to StartTime 
* DEV-30169: api alert evaluation
* DEV-16492 Support external alert evaluation 
* (no ticket) Upgrade to 8.4.0 
* DEV-18005 - add requestId in error message
* DEV-20400 - Grafana alerts evaluation - set 'accountsToSearch' query param

### CODE REVIEW

**Review by Jonathan-Eng — COMMENTED** [review]
as discussed. the pr is not ready yet for review

File: pkg/services/ngalert/api/api.go [hunk 1]
```diff
@@ -2,6 +2,7 @@ package api
 
 import (
 	"context"
+	"github.com/grafana/grafana/pkg/services/ngalert/schedule"
```
Comment (ohadza) [inline pkg/services/ngalert/api/api.go:5]: mark with `// LOGZ.IO GRAFANA CHANGE`

File: pkg/services/ngalert/api/api.go [hunk 2]
```diff
@@ -75,6 +76,7 @@ type API struct {
 	Tracer               tracing.Tracer
 	AppUrl               *url.URL
 	UpgradeService       migration.UpgradeService
+	Schedule             schedule.ScheduleService
```
Comment (ohadza) [inline pkg/services/ngalert/api/api.go:79]: mark with `// LOGZ.IO GRAFANA CHANGE`

File: pkg/services/ngalert/api/api_logzio_alerting.go [hunk 3]
```diff
@@ -0,0 +1,62 @@
...
+	OrgId int64 `json:"orgId"`
+}
+
+type AlertMigrationEmailNotification struct {
+	EmailAddress string `json:"address"`
+	ChannelUid   string `json:"channelUid"`
+}
+
+// NewLogzioAlertingApi creates a new LogzioAlertingApi instance
+func NewLogzioAlertingApi(service *LogzioAlertingService) *LogzioAlertingApi {
```
Comment (ohadza) [inline pkg/services/ngalert/api/api_logzio_alerting.go:33]: Didn't we say its redundant? (leave only the service)

File: pkg/services/ngalert/api/tooling/definitions/alertmanager.go [hunk 4]
```diff
@@ -4,6 +4,9 @@ import (
 	"context"
 	"encoding/json"
 	"fmt"
+
+	"github.com/grafana/grafana/pkg/services/ngalert/models"
+	// LOGZ.IO GRAFANA CHANGE :: End
```
Comment (ohadza) [inline pkg/services/ngalert/api/tooling/definitions/alertmanager.go:9]: where is the start :)

File: pkg/services/ngalert/schedule/schedule.go [hunk 5]
```diff
@@ -467,6 +551,8 @@ func (sch *schedule) ruleRoutine(grafanaCtx context.Context, key ngmodels.AlertR
 				attribute.Int64("results", int64(len(results))),
 			))
 		}
+
+		logger.Debug("RuleRoutine evaluation: processing evaluation results")
```
Comment (ohadza) [inline pkg/services/ngalert/schedule/schedule.go]: you should cleanup eventually all the logger.Debug(...) and comments ....

File: pkg/api/alerting.go [hunk 6]
```diff
@@ -9,6 +9,7 @@ import (
 
 	"github.com/grafana/grafana/pkg/api/dtos"
 	"github.com/grafana/grafana/pkg/api/response"
+	"github.com/grafana/grafana/pkg/models" // LOGZ.IO GRAFANA CHANGE :: DEV-17927 - add LogzIoHeaders
```
[the remark this answers is not in the document]
Comment (yasmin-tr) [reply pkg/api/alerting.go]: done

File: pkg/services/ngalert/api/tooling/definitions/alertmanager.go (hunk 4 above)
Comment (yasmin-tr) [reply pkg/services/ngalert/api/tooling/definitions/alertmanager.go]: probably some merge messed this up 😅  will fix it 👍

File: pkg/api/alerting.go [hunk 7]
```diff
@@ -179,7 +180,8 @@ func (hs *HTTPServer) AlertTest(c *contextmodel.ReqContext) response.Response {
 		return response.Error(400, "The dashboard needs to be saved at least once before you can test an alert rule", nil)
 	}
 
-	res, err := hs.AlertEngine.AlertTest(c.SignedInUser.GetOrgID(), dto.Dashboard, dto.PanelId, c.SignedInUser)
+	// LOGZ.IO GRAFANA CHANGE :: DEV-17927 - add LogzIoHeaders
+	res, err := hs.AlertEngine.AlertTest(c.SignedInUser.GetOrgID(), dto.Dashboard, dto.PanelId, c.SignedInUser, &models.LogzIoHeaders{RequestHeaders: c.Req.Header})
```
[the remark this answers is not in the document]
Comment (yasmin-tr) [reply pkg/api/alerting.go]: it is technically part of the logzio datasources support, but because a lot of the implementation for the logz evaluation includes this i put it in this PR.

i don't know if at that point it's worth the time of taking out of this PR for the next , because the next PR anyway will be based on that and making these changes can add up a day of work i don't see any value for it 😅 
in hindsight i would make the split if i knew about it from the beginning 😓

File: pkg/services/ngalert/api/api_logzio_alerting.go (hunk 3 above)
Comment (yasmin-tr) [reply pkg/services/ngalert/api/api_logzio_alerting.go]: this is the structure of all APIs in this project. there is a class of "XXXApi" that has the route to the endpoint, and the function to call the service class that contains the main logic.
i think what we said is that maybe the LogzioAlertingService can be redundant and we can put the logic directly in this class. but actually i think this structure has some sense in it.. also i'm pretty sure we'll want it also for later when more logic/apis will be added (or at least it gives the infra for it).

File: pkg/promlib/querydata/request.go [hunk 8]
```diff
@@ -110,27 +110,27 @@ func (s *QueryData) Execute(ctx context.Context, req *backend.QueryDataRequest)
...
 }
 
-func (s *QueryData) fetch(ctx context.Context, client *client.Client, q *models.Query, enablePrometheusDataplane bool) *backend.DataResponse {
+func (s *QueryData) fetch(ctx context.Context, client *client.Client, q *models.Query, enablePrometheusDataplane bool, headers map[string]string) *backend.DataResponse { // LOGZ.IO GRAFANA CHANGE :: DEV-43889 - Add headers for logzio datasources support
 	traceCtx, end := s.trace(ctx, q)
 	defer end()
 
 	logger := s.log.FromContext(traceCtx)
-	logger.Debug("Sending query", "start", q.Start, "end", q.End, "step", q.Step, "query", q.Expr)
+	logger.Debug("fetching query", "query", q, "headers", headers)
```
Comment (yasmin-tr) [inline pkg/promlib/querydata/request.go]: will fix this to be the original - for some reason didn't have the log so i added it , not sure how this happened

File: pkg/promlib/client/client_test.go [hunk 9]
```diff
@@ -94,7 +94,7 @@ func TestClient(t *testing.T) {
 				RangeQuery: true,
 				Step:       1 * time.Second,
 			}
-			res, err := client.QueryRange(context.Background(), req)
+			res, err := client.QueryRange(context.Background(), req, map[string]string{})
```
Comment (yasmin-tr) [inline pkg/promlib/client/client_test.go]: will add comment

File: pkg/promlib/querydata/request.go (hunk 8 above)
Comment (yasmin-tr) [inline pkg/promlib/querydata/request.go]: will fix this to be the original - for some reason didn't have the log so i added it , not sure how this happened
